tooling/SecretManager.VisualStudio/ProjectLocalSecretsManager.cs

```diff
@@ -0,0 +1,107 @@
+
+        private readonly IProjectProperties _projectProperties;
+
+        public ProjectLocalSecretsManager(IProjectProperties properties)
+        {
+            _projectProperties = properties;
+        }
+
+        public string SanitizeName(string name) => name;
+
+        public IReadOnlyCollection<char> GetInvalidCharactersFrom(string name) => Array.Empty<char>();
+
+        public Task AddSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+            => SetSecretAsync(name, value, cancellationToken);
+
+        public async Task SetSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            store.Set(name, value);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
+            }
+        }
+
+        public async Task<string> GetSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+            var store = await GetOrCreateStoreAsync();
+            return store[name];
+        }
+
+        public async Task<IReadOnlyCollection<string>> GetSecretNamesAsync(CancellationToken cancellationToken = default)
+        {
+            var store = await GetOrCreateStoreAsync();
+            return store.ReadOnlyKeys;
+        }
+
+        public async Task<IReadOnlyDictionary<string, string>> GetSecretsAsync(CancellationToken cancellationToken = default)
+        {
+            return await GetOrCreateStoreAsync();
+        }
+
+        public async Task<bool> RemoveSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            var result = store.Remove(name);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
+            }
+
+            return result;
+        }
+
+        private void EnsureKeyNameIsValue(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(nameof(name));
+            }
+        }
+
+        private async Task<SecretsStore> GetOrCreateStoreAsync()
+        {
+            var userSecretsId = await _projectProperties.GetEvaluatedPropertyValueAsync(UserSecretsPropertyName);
+
+            if (string.IsNullOrEmpty(userSecretsId))
+            {
+                userSecretsId = Guid.NewGuid().ToString();
+                await _projectProperties.SetPropertyValueAsync(UserSecretsPropertyName, userSecretsId);
+            }
+
+            return new SecretsStore(userSecretsId);
```

Comment:
Given users can open the secrets.json file in the VS editor, and in fact a command is exposed in VS to do just that, I think it appropriate to follow the best practice of first trying to open the file from the running document table if it is present there, and then updating that buffer with the change if it was opened from there -- only falling back on I/O if the document isn't already open. That gives the *integrated* development environment feeling instead of running an operation in the IDE that then pops a dialog saying "you made a change outside VS--would you like to reload the file?" to the user.

Reply:
Hmm, never heard of running document table. I'll look into this.